tests/unit/test_unique_any.cpp

```diff
@@ -0,0 +1,394 @@
+    EXPECT_EQ(typeid(std::string), anys.type());
+    EXPECT_EQ(typeid(std::vector<int>), anyv.type());
+    EXPECT_EQ(typeid(void), any0.type());
+
+    anyi.reset();
+    EXPECT_EQ(typeid(void), anyi.type());
+
+    anyi = std::true_type();
+    EXPECT_EQ(typeid(std::true_type), anyi.type());
+}
+
+TEST(unique_any, swap) {
+    using util::unique_any;
+    using util::any_cast;
+
+    unique_any any1(42);     // integer
+    unique_any any2(3.14);   // double
+
+    EXPECT_EQ(typeid(int),    any1.type());
+    EXPECT_EQ(typeid(double), any2.type());
+
+    any1.swap(any2);
+
+    EXPECT_EQ(any_cast<int>(any2), 42);
+    EXPECT_EQ(any_cast<double>(any1), 3.14);
+
+    EXPECT_EQ(typeid(double), any1.type());
+    EXPECT_EQ(typeid(int),    any2.type());
+
+    any1.swap(any2);
+
+    EXPECT_EQ(any_cast<double>(any2), 3.14);
+    EXPECT_EQ(any_cast<int>(any1), 42);
+
+    EXPECT_EQ(typeid(int),    any1.type());
+    EXPECT_EQ(typeid(double), any2.type());
+}
+
+TEST(unique_any, not_copy_constructable) {
+    using T = testing::nocopy<int>;
+
+    util::unique_any a(T(42));
+
+    auto& ref = util::any_cast<T&>(a);
+    EXPECT_EQ(ref, 42);
+    ref.value = 100;
+
+    EXPECT_EQ(util::any_cast<T&>(a).value, 100);
+
+    // the following will fail if uncommented, because we are requesting
+    // a copy of an non-copyable type.
+
+    //auto value = util::any_cast<T>(a);
+
+    // Test that we can move the contents of the unique_any.
+    // NOTE: it makes sense to sink a with std::move(a) instead
+    // of the following, because after such an assignment a will
+    // be in a moved from state, and enforcing std::move(a) it
+    // is made clearer in the calling code that a has been invalidated.
+    //     util::any_cast<T&&>(a)
+    T val(util::any_cast<T&&>(std::move(a)));
+    EXPECT_EQ(val.value, 100);
+}
+
+// test any_cast(unique_any*) and any_cast(const unique_any*)
+//   - these have different behavior to any_cast on reference types
+//   - are used by the any_cast on refernce types
+TEST(unique_any, any_cast_ptr) {
+    using util::unique_any;
+
+    // test that valid pointers are returned for int and std::string types
+
+    unique_any ai(42);
+    auto ptr_i = util::any_cast<int>(&ai);
+    EXPECT_EQ(*ptr_i, 42);
+
+    unique_any as(std::string("hello"));
+    auto ptr_s = util::any_cast<std::string>(&as);
+    EXPECT_EQ(*ptr_s, "hello");
+
+    // test that exceptions are thrown for invalid casts
+    EXPECT_EQ(util::any_cast<int>(&as), nullptr);
+    EXPECT_EQ(util::any_cast<std::string>(&ai), nullptr);
+    unique_any empty;
+    EXPECT_EQ(util::any_cast<int>(&empty), nullptr);
+    EXPECT_EQ(util::any_cast<int>((util::unique_any*)nullptr), nullptr);
+
+    // Check that constness of the returned pointer matches that the input.
+    {
+        unique_any a(42);
+        auto p = util::any_cast<int>(&a);
+        static_assert(std::is_same<int*, decltype(p)>::value,
+                "any_cast(any*) should not return const*");
```

Comment:
Carelessness.
More generally, which is the best approach?
Using `EXPECT_TRUE` has the benefit of not stopping compilation and running of the other tests if we run into a buggy compiler or if someone makes a breaking change.
But compile time checks give earlier and harder diagnosis of a problem.